compact.go

```diff
@@ -77,48 +84,59 @@ func (c Compact) Bytes() []byte {
 	return append([]byte{(topSixBits << 2) + 3}, b...)
 }
 
-func DecodeCompact(buffer *bytes.Buffer) (Compact, error) {
+func DecodeCompact[T BigNumbers](buffer *bytes.Buffer) (Compact[BigNumbers], error) {
 	decoder := Decoder{Reader: buffer}
 	result := make([]byte, 16)
 	b, err := decoder.DecodeByte()
 	if err != nil {
-		return Compact{}, err
+		return Compact[BigNumbers]{}, err
 	}
 	mode := b & 3
 	switch mode {
 	case 0:
-		return Compact(NewU128(big.NewInt(0).SetUint64(uint64(b >> 2)))), nil
+		switch reflect.TypeOf(*new(T)) {
+		case reflect.TypeOf(*new(U128)):
+			return Compact[BigNumbers]{Number: NewU128(big.NewInt(0).SetUint64(uint64(b >> 2)))}, nil
+		case reflect.TypeOf(*new(U64)):
+			return Compact[BigNumbers]{Number: NewU64(uint64(b >> 2))}, nil
+		case reflect.TypeOf(*new(U32)):
+			return Compact[BigNumbers]{Number: NewU32(uint32(b >> 2))}, nil
+		case reflect.TypeOf(*new(U8)):
+			return Compact[BigNumbers]{Number: NewU8(b >> 2)}, nil
+		default:
+			return Compact[BigNumbers]{Number: NewU128(big.NewInt(0).SetUint64(uint64(b >> 2)))}, nil
+		}
```

Comment:
Shouldn't these `Compact[BigNumbers]` be `Compact[T]`?

Reply:
If I make it as T, I'm having problems casting T to U128, U64 etc. For example, this line:
`return Compact[T]{Number: NewU128(big.NewInt(0).SetUint64(uint64(b >> 2)))}, nil`

Gives me the following error: **Cannot use 'NewU128(big.NewInt(0).SetUint64(uint64(b >> 2)))' (type U128) as the type T**

Reply:
There should be a prettier way to write this.
```golang
value := NewU64(uint64(b >> 2))
result := interface{}(value).(T)
return Compact[T]{result}, nil
```